File: semaphores/barrier.py
#! /usr/bin/env python

import threading
import time
import random

THREADS =   10
PHASES  =   20

""" Note we use mutex not only for incrementing and decrementing counter
    but also for printing, so the text don't get mangled in the console """

class Barrier(object):
    def __init__(self, n):
        self.n = n
        self.counter = 0
        self.mutex = threading.Lock()
        self.arrival = threading.Semaphore(1)
        self.departure = threading.Semaphore(0)

    def barrier(self):
        self.arrival.acquire()
        with self.mutex:
            self.counter += 1

        if self.counter < self.n:
            self.arrival.release()
        else:
            self.departure.release()

        self.departure.acquire()

        with self.mutex:
            self.counter -= 1

        if self.counter > 0:
            self.departure.release()
        else:
            self.arrival.release()



def thread(barrier):
    id = threading.current_thread().name
    print("Start thread {}".format(id))

    for i in range(PHASES):
        time.sleep(random.random())
        barrier.barrier()
        print("{} finished phase {}".format(id, i))

    barrier.barrier()
    print("Finished thread {}".format(id))

def main():
    threads = []
    barrier = Barrier(THREADS)

    for i in range(THREADS):
        # Create new threads
        t = threading.Thread(target=thread, args=(barrier,))
        threads.append(t)

    # Start all threads
    for t in threads:
        t.start()

    # Wait for all threads to complete
    for t in threads:
        t.join()



if __name__ == "__main__":
    main()

File: semaphores/Barrier.java
import java.util.Random;
import java.util.concurrent.Semaphore;
import java.util.concurrent.locks.ReentrantLock;

public class Barrier implements Runnable {
    static final int THREADS = 10;
    static final int PHASES  = 20;

    static volatile int arrived = 0, counter = 0;
    static ReentrantLock mutex = new ReentrantLock();
    static Semaphore arrival = new Semaphore(1);
    static Semaphore departure = new Semaphore(0);

    void barrier(int n) {
        try {
            arrival.acquire();
        } catch (InterruptedException e1) {}
        mutex.lock();
        counter++;
        mutex.unlock();
        if (counter < n) {
            arrival.release();
        } else {
            departure.release();
        }
        try {
            departure.acquire();
        } catch (InterruptedException e) {}
        mutex.lock();
        counter--;
        mutex.unlock();
        if (counter > 0) {
            departure.release();
        } else {
            arrival.release();
        }
    }


    @Override
    public void run() {
        long id = Thread.currentThread().getId();
        Random rnd = new Random();

        System.out.printf("Thread %d\n", id);

        for (int i = 0; i < PHASES; i++) {
            try {
                Thread.sleep(rnd.nextInt(1000));
            } catch (InterruptedException e) {}

            barrier(THREADS);
            System.out.printf("%d finished phase %d\n", id, i);
        }
        barrier(THREADS);
        System.out.printf("Finished thread %d\n", id);

    }

    public static void main(String[] args) throws InterruptedException {
        Thread[] threads = new Thread[THREADS];
        int i;

        Barrier c = new Barrier();

        for (i=0; i< THREADS; i++) {
            threads[i] = new Thread(c);
            threads[i].start();
        }

        for (i=0; i< THREADS; i++) {
            threads[i].join();
        }
    }
}
